cpu/sam0_common/periph/usbdev.c

```diff
@@ -0,0 +1,768 @@
+}
+
+static int _usbdev_ep_get(usbdev_ep_t *ep, usbopt_ep_t opt,
+                          void *value, size_t max_len)
+{
+    (void)max_len;
+    int res = -ENOTSUP;
+    switch (opt) {
+        case USBOPT_EP_STALL:
+            assert(max_len == sizeof(usbopt_enable_t));
+            *(usbopt_enable_t *)value = _ep_get_stall(ep);
+            res = sizeof(usbopt_enable_t);
+            break;
+        case USBOPT_EP_AVAILABLE:
+            assert(max_len == sizeof(size_t));
+            *(size_t *)value = _ep_get_available(ep);
+            res = sizeof(size_t);
+            break;
+        default:
+            DEBUG("sam_usb: Unhandled get call: 0x%x\n", opt);
+            break;
+    }
+    return res;
+}
+
+static int _usbdev_ep_set(usbdev_ep_t *ep, usbopt_ep_t opt,
+                          const void *value, size_t value_len)
+{
+    (void)value_len;
+    int res = -ENOTSUP;
+    switch (opt) {
+        case USBOPT_EP_ENABLE:
+            assert(value_len == sizeof(usbopt_enable_t));
+            if (*((usbopt_enable_t *)value)) {
+                _ep_enable(ep);
+                _usbdev_ep_init(ep);
+            }
+            else {
+                _ep_disable(ep);
+            }
+            res = sizeof(usbopt_enable_t);
+            break;
+        case USBOPT_EP_STALL:
+            assert(value_len == sizeof(usbopt_enable_t));
+            _ep_set_stall(ep, *(usbopt_enable_t *)value);
+            res = sizeof(usbopt_enable_t);
+            break;
+        case USBOPT_EP_READY:
+            assert(value_len == sizeof(usbopt_enable_t));
+            if (*((usbopt_enable_t *)value)) {
+                _ep_unready(ep);
+            }
+            else {
+                _usbdev_ep_ready(ep, 0);
+            }
+            res = sizeof(usbopt_enable_t);
+            break;
+        default:
+            DEBUG("sam_usb: Unhandled set call: 0x%x\n", opt);
+            break;
+    }
+    return res;
+}
+
+static int _ep_unready(usbdev_ep_t *ep)
+{
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+
+    if (ep->dir == USB_EP_DIR_IN) {
+        ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSCLR_BK1RDY;
+    }
+    else {
+        ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSSET_BK0RDY;
+    }
+    return 0;
+}
+
+static int _usbdev_ep_ready(usbdev_ep_t *ep, size_t len)
+{
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+
+    if (ep->dir == USB_EP_DIR_IN) {
+        _disable_ep_stall_in(ep_reg);
+        _bank_from_ep(ep)->PCKSIZE.bit.BYTE_COUNT = len;
+        ep_reg->EPSTATUSSET.reg = USB_DEVICE_EPSTATUSSET_BK1RDY;
+    }
+    else {
+        _disable_ep_stall_out(ep_reg);
+        ep_reg->EPSTATUSCLR.reg = USB_DEVICE_EPSTATUSCLR_BK0RDY;
+    }
+    return 0;
+}
+
+/**
+ * Endpoint event handler
+ *
+ * Calls the endpoint callback to report the event to the USB stack
+ */
+static void _usbdev_ep_esr(usbdev_ep_t *ep)
+{
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+    signed event = -1;
```

Comment:
int32 or int16 ? I'm sure this one will be troublesome outside of ARM.

Reply:
But this driver is closely tied to an ARM based core…

I can change it to `int32_t` if you like

Reply:
But same as [this comment](https://github.com/RIOT-OS/RIOT/pull/11213#discussion_r267553921), there is not really a need for a specific variable size here, so I'd rather give the compiler more freedom for possible optimizations.